lib/plug/conn.ex

```diff
@@ -1328,50 +1340,148 @@ defmodule Plug.Conn do
+  or consider explicitly escaping the cookie value by using a function
+  such as `Base.encode64(value, padding: false)` when writing and
+  `Base.decode64(encoded, padding: false)` when reading the cookie.
+  It is important for padding to be disabled since `=` is not a valid
+  character in cookie values.
+
+  ## Signing and encrypting cookies
+
+  This function allows you to automatically sign and encrypt cookies.
+  When signing or encryption is enabled, then any Elixir value can be
+  stored in the cookie (except anonymous functions for security reasons).
+  Once a value is signed or encrypted, you must also call `fetch_cookie/2`
+  with the cookies that are either signed or encrypted.
+
+  To sign, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, sign: true)
+
+  and then:
+
+      fetch_cookie(conn, signed: ~w(my-cookie))
+
+  To encrypt, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, encrypt: true)
+
+  and then:
+
+      fetch_cookie(conn, encrypted: ~w(my-cookie))
+
+  By default a signed or encrypted cookie is only valid for a day, unless
+  a `:max_age` is specified.
 
   ## Options
 
     * `:domain` - the domain the cookie applies to
     * `:max_age` - the cookie max-age, in seconds. Providing a value for this
-      option will set both the _max-age_ and _expires_ cookie attributes
+      option will set both the _max-age_ and _expires_ cookie attributes.
     * `:path` - the path the cookie applies to
     * `:http_only` - when `false`, the cookie is accessible beyond HTTP
     * `:secure` - if the cookie must be sent only over https. Defaults
       to true when the connection is HTTPS
     * `:extra` - string to append to cookie. Use this to take advantage of
       non-standard cookie attributes.
+    * `:signed` - when true, signs the
 
   """
   @spec put_resp_cookie(t, binary, binary, Keyword.t()) :: t
   def put_resp_cookie(%Conn{} = conn, key, value, opts \\ [])
-      when is_binary(key) and is_binary(value) and is_list(opts) do
+      when is_binary(key) and is_list(opts) do
     %{resp_cookies: resp_cookies, scheme: scheme} = conn
-    cookie = [{:value, value} | opts] |> :maps.from_list() |> maybe_secure_cookie(scheme)
+    {to_send_value, opts} = maybe_sign_or_encrypt_cookie(conn, key, value, opts)
+    cookie = [{:value, to_send_value} | opts] |> Map.new() |> maybe_secure_cookie(scheme)
     resp_cookies = Map.put(resp_cookies, key, cookie)
     update_cookies(%{conn | resp_cookies: resp_cookies}, &Map.put(&1, key, value))
   end
 
+  defp maybe_sign_or_encrypt_cookie(conn, key, value, opts) do
+    {sign?, opts} = Keyword.pop(opts, :sign, false)
+    {encrypt?, opts} = Keyword.pop(opts, :encrypt, false)
+
+    case {sign?, encrypt?} do
+      {true, true} ->
+        raise ArgumentError, "only :sign or :encrypt may be given at the same time"
+
+      {true, false} ->
+        {Plug.Crypto.sign(conn.secret_key_base, key, value, signed_at(opts)), opts}
```

Comment:
@GriffinMB @voltone  - I am using the cookie key as the salt. Is this a good or bad idea?

Reply:
Hmm, I am not a cryptographer, but as long as the cookie names are not under external control it should be ok, I think. Give me a day to think about it...

Reply:
Since it's only really used as part of key generation, it should be fine as long as it isn't user controllable. Though, users should be aware that they should not re-use that salt, and we need to be careful not to re-use the salt incidentally anywhere else.

Reply:
Right, that's the main concern here: the fact that a key is derived from the global secret_key_base with a salt that's implicitly selected by this function. If the user derived a key using the same salt in a different context, and happens to use it in a way that can act as a signing oracle, then an attacker could produce their own cookie values. It might not be likely, but it is impossible to rule out

Reply:
I now append a "_cookie" suffix to all cookie names. Let me know what you think.

Reply:
That sounds good to me. Probably still worth writing down in the docs somewhere, but it does seem unlikely that an unintended collision would occur.

Reply:
It has been documented, thanks!